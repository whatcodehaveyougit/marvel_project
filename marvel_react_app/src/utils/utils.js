import MD5 from "crypto-js/md5";

const getHash = (ts, privateKey, publicKey) => {
    return MD5(ts + privateKey + publicKey).toString();
};

const fetchData = async () => {

    const baseUrl = `${process.env.REACT_APP_API_URL}/characters`
    const ts = Date.now().toString()
    const privateKey = process.env.REACT_APP_PRIVATE_KEY;
    const publicKey = process.env.REACT_APP_PUBLIC_KEY;
    const hash = getHash(ts, privateKey, publicKey);
    const url = `${baseUrl}?ts=${ts}&apikey=${publicKey}&hash=${hash}`

    // console.log('priv key: ' + privateKey);
    // console.log('public key: ' + publicKey);

    try {
        let response = await fetch (url);
        const data = await response.json()
        return data;
    } catch ( err ) {
        return err;
    }

}

export { fetchData };