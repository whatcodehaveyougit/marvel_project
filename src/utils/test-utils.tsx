import { render } from "@testing-library/react";
import
import { TranslationProvider } from "my-i18n-lib";
import defaultStrings from "i18n/en-x-default";

const AllTheProviders = ({ children }) => {
  return (
    <Provider theme="light">
      <TranslationProvider messages={defaultStrings}>
        {children}
      </TranslationProvider>
    </Provider>
  );
};

const customRender = (ui: any, options: any) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
