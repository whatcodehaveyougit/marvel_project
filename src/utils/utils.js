import MD5 from "crypto-js/md5";

const getHash = (ts, privateKey, publicKey) => {
    return MD5(ts + privateKey + publicKey).toString();
};

const fetchData = async ( apiRoute ) => {

    const baseUrl = `${process.env.REACT_APP_API_URL}${apiRoute}`
    const ts = Date.now().toString()
    const privateKey = process.env.REACT_APP_PRIVATE_KEY;
    const publicKey = process.env.REACT_APP_PUBLIC_KEY;
    const hash = getHash(ts, privateKey, publicKey);
    const url = `${baseUrl}?ts=${ts}&apikey=${publicKey}&hash=${hash}`

    try {
        let response = await fetch (url);
        const data = await response.json()
        return data;
    } catch ( err ) {
        return err;
    }

}

export { fetchData };