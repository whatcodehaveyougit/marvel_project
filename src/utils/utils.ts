import MD5 from "crypto-js/md5";
import { TCharacter } from "../types/types";
import axios from "axios";

const getHash = (
  ts: string,
  privateKey: string | undefined,
  publicKey: string | undefined
) => {
  return MD5(ts + privateKey + publicKey).toString();
};

// ToDo - The return type of this function ?
const fetchData = async (apiRoute: string) => {
  const baseUrl = `${process.env.REACT_APP_API_URL}${apiRoute}`;
  const ts = Date.now().toString();
  const privateKey = process.env.REACT_APP_PRIVATE_KEY;
  const publicKey = process.env.REACT_APP_PUBLIC_KEY;
  const hash = getHash(ts, privateKey, publicKey);
  const url = `${baseUrl}?ts=${ts}&apikey=${publicKey}&hash=${hash}`;
  try {
    let response = await axios.get(url);
    return response["data"]["data"]["results"];
  } catch (err: any | unknown) {
    throw Error(err);
  }
};

const generateBackgroundImageUrl = (
  characterData: TCharacter | null
): string => {
  if (characterData) {
    if (characterData["thumbnail"]) {
      const backgroundImageUrlConcatenated =
        characterData["thumbnail"]["path"] +
        "." +
        characterData["thumbnail"]["extension"];
      return backgroundImageUrlConcatenated;
    }
  }
  return "";
};

function showFirstNCharacters(string: string, n: number | null): string {
  if (n) {
    return string.slice(0, n) + "...";
  }
  return string;
}

export { fetchData, generateBackgroundImageUrl, showFirstNCharacters };
