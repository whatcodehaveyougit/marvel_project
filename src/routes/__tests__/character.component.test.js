import { render, cleanup, waitFor } from '@testing-library/react'
import { MemoryRouter as Router } from "react-router-dom";
import CharacterPage from '../character/character.component';
import { fetchData } from '../../utils/utils';

afterEach(cleanup)

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    characterid: '123',
  }),
}));

// If we don't have the CharactersData we do the call, here is response
const characterMockedAPIResponse = {
  data: {
    results: {
          id: '123',
          name: 'The Hulk',
          thumbnail: {
            path: 'this-it-the-path2',
            extension: 'jpg'
          },
          description: 'He is a big monstor'
      }
  }
}

const characterComicsMockedAPIResponse = {
  data: {
    results: [
      {
        id: 1,
        title: 'A title',
        description: 'Hello World'
      },
      {
        id: 2,
        title: 'The title',
        description: 'Hello World'
      },
    ]
  }
}

  // We pass the characters API Response if we have the data
  const charactersData = [
    {
      id: '890',
      name: 'The Man',
      thumbnail: {
        path: 'this-it-the-path',
        extension: 'jpg'
      },
      description: 'He is a big monstor'
    },
    {
      id: '123',
      name: 'The Hulk',
      thumbnail: {
        path: 'this-it-the-path2',
        extension: 'jpg'
      },
      description: 'He is a big monstor'
    }
  ]

const MockCharacterPage = () => {
  return (
    <Router>
      <CharacterPage
      charactersData={charactersData} />
    </Router>
  )
}

jest.mock('../../utils/utils', () => ({
  fetchData: jest.fn((params) => {
    if (params === '/characters/123/comics'){
      return characterComicsMockedAPIResponse
    } else if ( params === '/characters/123'){
      return characterMockedAPIResponse;
    }
  })
}))


describe('CharacterPage Tests', function() {



  it('CharacterPage Snapshot with data given', async () => {

    const {container }  = render(
      <MockCharacterPage />
    )
    await waitFor(() => {
      expect(container).toMatchSnapshot()
    })
  })


  it('CharacterPage Snapshot with no data given', async () => {

    const {container }  = render(
      <Router>
        <CharacterPage
          charactersData={[]} />
      </Router>
    )
    await waitFor(() => {
      // Sigurd - WHY called 3 times???
      expect(fetchData).toHaveBeenCalledTimes(3);
      expect(container).toMatchSnapshot()
    })
  })


})




