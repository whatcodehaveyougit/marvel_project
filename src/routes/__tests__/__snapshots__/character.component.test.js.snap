// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterPage Tests CharacterPage Snapshot with data given 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root character-page css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="thumbnail-image"
      style="background-image: url(this-it-the-path2.jpg);"
    >
      <div
        class="MuiCardContent-root character-information css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiGrid-root character-page-headings css-vj1n65-MuiGrid-root"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-47suzj-MuiTypography-root"
          >
            The Hulk
          </h3>
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
          >
            List of comics for this character:
          </h5>
        </div>
        <div
          class="MuiGrid-root css-vj1n65-MuiGrid-root"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CharacterPage Tests CharacterPage Snapshot with no data given 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root character-page css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="thumbnail-image"
      style="background-image: url(undefined);"
    >
      <div
        class="MuiCardContent-root character-information css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiGrid-root character-page-headings css-vj1n65-MuiGrid-root"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-47suzj-MuiTypography-root"
          />
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
          >
            List of comics for this character:
          </h5>
        </div>
        <div
          class="MuiGrid-root css-vj1n65-MuiGrid-root"
        />
      </div>
    </div>
  </div>
</div>
`;
